src/sync/mutex.rs

```diff
@@ -24,78 +24,61 @@ pub struct MutexGuard<'a, T: ?Sized> {
 #[derive(Debug)]
 struct MutexState {
     holder: Option<TaskId>,
-    waiters: TaskSet,
-    clock: VectorClock,
 }
 
 impl<T> Mutex<T> {
     /// Creates a new mutex in an unlocked state ready for use.
     pub const fn new(value: T) -> Self {
         let state = MutexState {
             holder: None,
-            waiters: TaskSet::new(),
-            clock: VectorClock::new(),
         };
-
         Self {
-            inner: std::sync::Mutex::new(value),
             state: RefCell::new(state),
+            semaphore: BatchSemaphore::const_new(1, Fairness::Unfair),
+            inner: std::sync::Mutex::new(value),
         }
     }
 }
 
 impl<T: ?Sized> Mutex<T> {
     /// Acquires a mutex, blocking the current thread until it is able to do so.
     pub fn lock(&self) -> LockResult<MutexGuard<'_, T>> {
-        let me = ExecutionState::me();
+        let me = current::me();
 
         let mut state = self.state.borrow_mut();
-        trace!(holder=?state.holder, waiters=?state.waiters, "waiting to acquire mutex {:p}", self);
+        trace!(holder=?state.holder, semaphore=?self.semaphore, "waiting to acquire mutex {:p}", self);
+        drop(state);
 
-        // If the lock is already held, then we are blocked
-        if let Some(holder) = state.holder {
-            if holder == me {
-                panic!("deadlock! task {:?} tried to acquire a Mutex it already holds", me);
-            }
+        // Try to acquire without blocking first. If we succeed, we are done
+        // and can move on to acquiring the inner mutex. This is a yield point.
+        match self.semaphore.try_acquire(1) {
+            Ok(()) => (),
+            Err(TryAcquireError::Closed) => unreachable!(),
+            Err(_) => {
+                // If the lock is already held, then we are blocked. This can
+                // happen either because the lock is held by another thread,
+                // or because the lock is held by the current thread. For the
+                // latter case, we check the state to report a more precise
+                // error message.
+                state = self.state.borrow_mut();
+                if let Some(holder) = state.holder {
```

Comment:
If this is `None`, then we have a bug in the `Mutex` – swap to `state.holder.expect(...)` ?

Reply:
I don't think this is the case: `state.holder` may still be `None` because there is a yield between the acquiry of a permit from the semaphore and the actual state update (the two are not performed in a single atomic step).

However, the check is still sufficient here because in the case of a thread attempting to re-acquire a lock it already holds, it must have returned from the first `lock` call by the time the second one is checking the state.